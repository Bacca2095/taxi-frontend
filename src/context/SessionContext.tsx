/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React, { createContext, useEffect, useRef, useState } from 'react';
import { SessionState } from 'context/models/SessionState';
import * as sessionService from '../services/sessionService';

export const useStateContainer = (
  initialState: SessionState = { document: '', sessionId: '' },
) => {
  const [document, setDocument] = useState(initialState.document);
  const [sessionId, setSessionId] = useState(initialState.sessionId);
  const documentRef = useRef(false);

  const clearSession = () => {
    sessionService.deleteSession();
  };

  useEffect(() => {
    if (!documentRef.current) {
      documentRef.current = true;
      return;
    }
    if (!document) {
      return;
    }
    setSessionId(sessionService.createSession(document));
  }, [document]);

  return {
    data: { document, sessionId },
    mutations: { setDocument, clearSession },
  };
};

export const SessionContext = createContext<
  ReturnType<typeof useStateContainer>
>({} as never);

export const SessionProvider: React.FC = ({ children }) => {
  const contextValue = useStateContainer(sessionService.recoverSession());
  return (
    <SessionContext.Provider value={contextValue}>
      {children}
    </SessionContext.Provider>
  );
};
