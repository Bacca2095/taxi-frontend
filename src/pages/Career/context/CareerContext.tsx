/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import React, {
  createContext,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react';
import { SessionContext } from 'context/SessionContext';
import * as careerService from '../services/careerService';
import { CareerModel } from '../models/CareerModel';

export interface CareerState {
  careers?: CareerModel[];
}

export const useStateContainer = (
  initialState: CareerState = {},
  sessionContext: typeof SessionContext,
) => {
  const {
    data: { sessionId },
  } = useContext(sessionContext);
  const [allCareers, setAllCareers] = useState(initialState.careers || []);
  const [currentCareerId, setCurrentCareerId] = useState(0);
  const [errorOnDelete, setErrorOnDelete] = useState(false);
  const [errorOnCreate, setErrorOnCreate] = useState(false);
  const [updateList, setUpdateList] = useState(false);
  const careerRef = useRef(false);

  const deleteCareer = () => {
    careerService.deleteCareer(currentCareerId).then((res) => {
      if (res) {
        setErrorOnDelete(false);
      } else {
        setErrorOnDelete(true);
      }
      setUpdateList(!updateList);
    });
  };

  const createCareer = (career: CareerModel) => {
    careerService.createCareer(career).then((res) => {
      if (res) {
        setErrorOnCreate(false);
      } else {
        setErrorOnCreate(true);
      }

      setUpdateList(!updateList);
    });
  };

  useEffect(() => {
    if (!careerRef.current) {
      careerRef.current = true;
      return;
    }
    if (!document) {
      return;
    }
    careerService
      .listCareer(sessionId?.split(':::')[1])
      .then((career) => setAllCareers(career));
  }, [sessionId, updateList]);

  return {
    data: { allCareers, currentCareerId, errorOnDelete, errorOnCreate },
    mutations: {
      setCurrentCareerId,
      deleteCareer,
      createCareer,
    },
  };
};

export const CareerContext = createContext<
  ReturnType<typeof useStateContainer>
>({} as never);

export interface CareerProviderProps {
  initialState?: CareerState;
}
export const CareerProvider: React.FC<CareerProviderProps> = ({
  children,
  initialState = {},
}) => {
  const contextValue = useStateContainer(initialState, SessionContext);
  return (
    <CareerContext.Provider value={contextValue}>
      {children}
    </CareerContext.Provider>
  );
};
