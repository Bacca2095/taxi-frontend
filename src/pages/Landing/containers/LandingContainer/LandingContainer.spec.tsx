import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { SessionProvider } from 'context/SessionContext';
import { LandingContainer } from '.';

describe('Landing Container tests', () => {
  let onContinue: jest.Mock;

  beforeEach(() => {
    onContinue = jest.fn();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <SessionProvider>
        <LandingContainer />
      </SessionProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
