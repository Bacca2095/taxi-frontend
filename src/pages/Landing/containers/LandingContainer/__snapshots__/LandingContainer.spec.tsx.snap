// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing Container tests should match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-container-2 MuiGrid-container MuiGrid-justify-content-xs-center"
  >
    <div
      class="MuiGrid-root makeStyles-containerItem-3 MuiGrid-container MuiGrid-item MuiGrid-align-content-xs-center MuiGrid-justify-content-xs-center MuiGrid-grid-xs-10 MuiGrid-grid-md-6"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 MuiGrid-grid-md-6"
      >
        <div
          class="MuiPaper-root MuiCard-root makeStyles-card-1 MuiPaper-elevation6 MuiPaper-rounded"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-align-content-xs-center MuiGrid-justify-content-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-10"
            >
              <p
                class="MuiTypography-root makeStyles-title-4 MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
              >
                Bienvenido a Taxi APP
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
              >
                Ingrese su documento para consultar y solicitar nuevas carreras
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-10"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-input-5 MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink="false"
                >
                  Numero de documento
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    data-testid="input-login-document"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-9"
                    >
                      <span>
                        Numero de documento
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
                data-testid="login"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Continuar
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
